### ISSUE TITLE
10.3.1 fails to compile on Ubuntu 17.10

### ISSUE THREAD

**Issue by fossfreedom** [budgie-desktop/budgie-desktop#951]
v10.3.1 budgie-desktop

build fails on all architectures

build log - https://launchpad.net/ubuntu/+source/budgie-desktop/10.3.1-2/+build/12735601

I think it is due to this bit in the build log (searched for "Failed:")

```
[34/221] rm -f src/imports/natray/libnatray.a && gcc-ar csrD src/imports/natray/libnatray.a 'src/imports/natray/natray@sta/na-tray.c.o' 'src/imports/natray/natray@sta/na-tray-child.c.o' 'src/imports/natray/natray@sta/na-tray-manager.c.o' 'src/imports/natray/natray@sta/fixedtip.c.o'
[35/221] valac -C -lm --pkg gtk+-3.0 --pkg libpeas-1.0 --pkg libpeas-1.0 --pkg gtk+-3.0 -d src/lib/toplevel@sta --library=toplevel -H src/lib/toplevel.h --use-header --vapi ../toplevel.vapi --pkg budgie-1.0 --pkg gio-unix-2.0 --pkg gtk+-3.0 --pkg libpeas-1.0 --vapidir /<<PKGBUILDDIR>>/src/plugin -H BudgieToplevel.h ../../src/lib/animation.vala ../../src/lib/toplevel.vala ../../src/lib/shadow.vala ../../src/lib/manager.vala
FAILED: src/lib/toplevel@sta/animation.c src/lib/toplevel@sta/toplevel.c src/lib/toplevel@sta/shadow.c src/lib/toplevel@sta/manager.c src/lib/toplevel.h src/lib/toplevel.vapi 
valac -C -lm --pkg gtk+-3.0 --pkg libpeas-1.0 --pkg libpeas-1.0 --pkg gtk+-3.0 -d src/lib/toplevel@sta --library=toplevel -H src/lib/toplevel.h --use-header --vapi ../toplevel.vapi --pkg budgie-1.0 --pkg gio-unix-2.0 --pkg gtk+-3.0 --pkg libpeas-1.0 --vapidir /<<PKGBUILDDIR>>/src/plugin -H BudgieToplevel.h ../../src/lib/animation.vala ../../src/lib/toplevel.vala ../../src/lib/shadow.vala ../../src/lib/manager.vala
Unknown option -lm
Run 'valac --help' to see a full list of available command line options.
```

Thoughts?

**Comment by ikeydoherty** [comment]
Oh weird. So we're passing the link flag to vala and not the C compiler

**Comment by fossfreedom** [comment]
ok - just trying to pin this down - Ubuntu has uplifted valac to 0.36.3 - what is Solus using @ikeydoherty ?

**Comment by ikeydoherty** [comment]
0.36.3

**Comment by fossfreedom** [comment]
hmm. meson?  ubuntu is now using 0.41.0

**Comment by ikeydoherty** [comment]
0.40.1 - interesting

**Comment by fossfreedom** [comment]
Think this is the issue. Looks like a regression https://github.com/mesonbuild/meson/issues/1939

**Comment by ikeydoherty** [comment]
Just shoved new meson into Solus to see the problem https://dev.solus-project.com/R2067:d6ceda841455e5cd6842055cdf78bb1a6bc4e682

**Comment by ikeydoherty** [comment]
It breaks further along too with headers being trashed, flags going the wrong way, etc. This is an upstream Meson issue and I'd recommend that Ubuntu rolls back to the working release (as I'm about to do). Literally nothing I can do in Ubuntu. Apparently Meson and Vala are spiritually aligned in breaking on updates.

**Comment by Vistaus** [comment]
Even with Meson 0.40.1 compiling fails:

```
> meson --version
0.40.1
[12:21:22] vistaus@vistaus-solus /media/MicroSD/Apps/budgie-desktop/build (0) 
> cd ..
[12:21:25] vistaus@vistaus-solus /media/MicroSD/Apps/budgie-desktop (0) 
> rm -rf build
[12:21:26] vistaus@vistaus-solus /media/MicroSD/Apps/budgie-desktop (0) 
> mkdir build
[12:21:29] vistaus@vistaus-solus /media/MicroSD/Apps/budgie-desktop (0) 
> cd build/
[12:21:30] vistaus@vistaus-solus /media/MicroSD/Apps/budgie-desktop/build (0) 
> meson .. --prefix=/usr
The Meson build system
Version: 0.40.1
Source dir: /media/MicroSD/Apps/budgie-desktop
Build dir: /media/MicroSD/Apps/budgie-desktop/build
Build type: native build
Project name: budgie-desktop
Native c compiler: cc (gcc 7.1.0)
Appending CFLAGS from environment: '-g2 -O3 -pipe -fPIC -Wformat -Wformat-security -fno-omit-frame-pointer -fexceptions -D_FORTIFY_SOURCE=2 -fstack-protector --param ssp-buffer-size=32 -fasynchronous-unwind-tables -ftree-vectorize -feliminate-unused-debug-types -Wall -Wno-error -Wp,-D_REENTRANT -march=native'
Appending LDFLAGS from environment: '-Wl,-z,relro -Wl,-z,now'
Native vala compiler: valac (valac 0.36.3)
Appending LDFLAGS from environment: '-Wl,-z,relro -Wl,-z,now'
Build machine cpu family: x86_64
Build machine cpu: x86_64
Program meson_post_install.sh found: YES (/media/MicroSD/Apps/budgie-desktop/meson_post_install.sh)
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 298, in run
    app.generate()
  File "/usr/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 180, in generate
    intr.run()
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreter.py", line 2504, in run
    super().run()
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 125, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 146, in evaluate_codeblock
    raise e
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 140, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 153, in evaluate_statement
    return self.assignment(cur)
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 566, in assignment
    value = self.evaluate_statement(node.value)
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 151, in evaluate_statement
    return self.function_call(cur)
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreterbase.py", line 372, in function_call
    return self.funcs[func_name](node, self.flatten(posargs), kwargs)
  File "/usr/lib/python3.5/site-packages/mesonbuild/interpreter.py", line 1861, in func_dependency
    identifier = dependencies.get_dep_identifier(name, kwargs)
TypeError: get_dep_identifier() missing 1 required positional argument: 'want_cross'
[12:21:35] vistaus@vistaus-solus /media/MicroSD/Apps/budgie-desktop/build (1) 
> 
```

Using Solus, fully up-to-date and Meson from the default Solus repository (stable).

**Comment by ikeydoherty** [comment]
It built fine for me on Solus doing:
```
 meson build --prefix=/usr --sysconfdir=/etc --libdir=/usr/lib64  --buildtype plain
```

**Comment by Vistaus** [comment]
Dunno what went wrong, but I deleted /usr/lib/python3.5/site-packages/mesonbuild/ and reinstalled Meson and now it does build fine.

**Comment by ikeydoherty** [comment]
Looks like you had an old manual install of meson

**Comment by Vistaus** [comment]
Btw, seems like this is fixed upstream now: https://github.com/mesonbuild/meson/issues/1939
So all we have to do is wait for the next Meson release (or compile Meson from git).

**Comment by fossfreedom** [comment]
@Vistaus we have fixed meson on 17.10 to allow the compilation of budgie desktop v10.3.1 ... This version is in artful dailies.

Note. Neither this version of meson not the upstream version of meson will compile the current git version of budgie desktop .  That is a separate issue and shouldn't be tagged on this correctly closed thread.
